File: DCC2024_Artur/January/Day8_Range_Sum_of_BST.cpp
/*
 https://leetcode.com/problems/range-sum-of-bst/?envType=daily-question&envId=2024-01-08
*/

#include <iostream>
#include <vector>

using namespace std;


struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high)
    {
        return postorder(root, low, high);
    }


    int postorder(TreeNode *node, int low, int high)
    {
        int res = 0;
        if( node->left )
            res += postorder(node->left, low, high);
        if( node->right )
            res += postorder(node->right, low, high);
        if( low <= node->val && node->val <= high )
            res += node->val;

        return res;
    }
};


int main()
{
    Solution solution;

    auto* node1 = new TreeNode(1);
    auto* node6 = new TreeNode(6);
    auto* node3 = new TreeNode(3, node1, nullptr);
    auto* node7 = new TreeNode(7, node6, nullptr);
    auto* node13 = new TreeNode(13);
    auto* node18 = new TreeNode(18);
    auto* node5 = new TreeNode(5, node3, node7);
    auto* node15 = new TreeNode(15, node13, node18);
    auto* node10 = new TreeNode(10, node5, node15);

    cout << solution.rangeSumBST(node10, 10, 18) << endl;

    return 0;
}

/*
 Main idea:
 Обходимо дерева у зворотньому порядку(ЛПК), на кожній рекурсивній
 ітерації повертається сума значень лівого, правого сина, а також
 кореня, якщо його значення входить в даний інтервал.
*/

File: DCC2024_Artur/January/Day8_Range_Sum_of_BST.py
"""
https://leetcode.com/problems/range-sum-of-bst/?envType=daily-question&envId=2024-01-08
"""

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        return self.postorder(root, low, high)
        
    def postorder(self, root: TreeNode, low: int, high: int) -> int:
        res = 0
        if root.left:
            res += self.postorder(root.left, low, high)
        if root.right:
            res += self.postorder(root.right, low, high)
        if low <= root.val <= high:
            res += root.val
        
        return res
        
        
        
        
        
if (__name__ == '__main__'):
    solution = Solution()

    node1 = TreeNode(1)
    node6 = TreeNode(6)
    node3 = TreeNode(3, node1)
    node7 = TreeNode(7, node6)
    node13 = TreeNode(13)
    node18 = TreeNode(18)
    node5 = TreeNode(5, node3, node7)
    node15 = TreeNode(15, node13, node18)
    node10 = TreeNode(10, node5, node15)
    
    print(solution.rangeSumBST(node10, 1, 18))